### DISCUSSION TITLE
Should functions check to see if boolean inputs are actually boolean?

### DISCUSSION THREAD

**Discussion by williamhobbs** [pvlib/pvlib-python#2320] [General]
I noticed that passing
```
backtracking = 'false'
```
in `pvlib.tracking.singleaxis()` results in backtracking being included. Should this, and other functions, check to make sure that boolean inputs are actually boolean and not a string that is not being interpreted as the user might have expected? 

I'm guessing this is a general Python best practices question, but I honestly don't know.

**Comment by cwhanse** [comment]
We ~~should~~ could restructure that check. Any "truthy" value (string, non-zero number, etc.) will be interpreted as `True`.

**Reply by williamhobbs** [reply]
Thanks, @cwhanse. It seems like several other functions handle boolean inputs the same way. Here are two, but I think there are more:

https://pvlib-python.readthedocs.io/en/stable/_modules/pvlib/clearsky.html#lookup_linke_turbidity has 
```python
    if interp_turbidity:
        linke_turbidity = _interpolate_turbidity(lts, time)
    else:
        months = tools._pandas_to_utc(time).month - 1
        linke_turbidity = pd.Series(lts[months], index=time)
```
_delta_kt_prime_dirint in pvlib.irradiance:
```python
    if use_delta_kt_prime:
        # Perez eqn 2
        kt_next = kt_prime.shift(-1)
...
```

**Reply by wholmgren** [reply]
I don't see the problem here... in Python `'its_a_duck'` evaluates to `True`. I don't think pvlib should fight python semantics.

**Reply by cwhanse** [reply]
I'm in the minority. My perspective is that a user will add a `backtrack` value only when they want behavior other than the default `backtrack=True`. Unless that value is `False` they will get the default behavior. My idea is that we restructure the check so that the backtrack code is only executed if `backtrack=True` and not executed otherwise.

**Reply by echedey-ls** [reply]
I understand your point @cwhanse . But if the problem is exclusively with this function, then the error is relatively easy to debug with plots (and also there many examples using `backtrack=False`). IDEs should also help with that. At least in VSCode (sorry for spamming again) with Pylance extension you get the option to opt-in for type checking:
![image](https://github.com/user-attachments/assets/c5f78268-1de0-482d-afb6-f36ba86782a9)

And that's by default, I don't need to specify explicitly `backtrack: bool = True` in `tracking.singleaxis` signature.

**Reply by williamhobbs** [reply]
> I don't think pvlib should fight python semantics.

@wholmgren, I thought this might be the case, i just didn't know what standard Python semantics were in this situation.

**Reply by mikofski** [reply]
According to the docs it is unpythonic to “look before you leap” and pythonic to follow “easier to ask forgiveness than permission”. This is explained in Microsoft developer Brett Cannon’s blogpost [EAFP vs LBYL](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/).

**Reply by cwhanse** [reply]
All I'm suggesting is that we could consider replacing

```
if backtrack:
```

with 

```
if backtrack is True:
```

That would probably alarm the linter, but it would restrict when backtracking code is executed to only `backtrack=True`.

In any case, I'm not invested in making a change here, only wanting to help pvlib be easier for new users.

**Reply by adriesse** [reply]
In my own code I have a habit of using 0 and 1 to set boolean parameters, but I have noticed it doesn't always work with other packages. I have a very slight preference for leaving it the way it is.

**Comment by adriesse** [comment]
Is there perhaps an inherent benefit in having default values be `False` or `None`?

**Reply by echedey-ls** [reply]
You are certainly right @adriesse ! 
In [my comment up there](https://github.com/pvlib/pvlib-python/discussions/2320#discussioncomment-11452689) my IDE/static code checker detects it should be a `bool` value because the default is `False`. If it had been `None` then the inference would be of type `Any`, and no warning would have been emitted.
